src/s3/commands/downloadFileAs.ts

```diff
@@ -163,3 +169,150 @@ async function promptForSaveLocation(
+        if (!fs.existsSync(dir)) {
+            fs.mkdirSync(dir)
+            saveLocation = vscode.Uri.file(dir)
+        }
+    } else {
+        // since we cannot control which nodes are selected and passed here, filter only S3FileNodes
+        allNodes = allNodes.filter(node => node instanceof S3FileNode)
+        files = allNodes?.map(fileNode => {
+            return { bucket: fileNode.bucket, ...fileNode.file }
+        })
+    }
+
+    let failed = await downloadAllFiles(files, saveLocation, node.s3)
+
+    showOutputMessage(
+        localize(
+            'AWS.s3.downloadFile.complete',
+            'Downloaded {0}/{1} files',
+            files.length - failed.length,
+            files.length
+        ),
+        outputChannel
+    )
+
+    while (failed.length > 0) {
+        const failedKeys = failed.map(file => file.key)
+        getLogger().error(`List of requests failed to download:\n${failedKeys.toString().split(',').join('\n')}`)
+
+        if (failed.length > 5) {
+            showOutputMessage(
+                localize(
+                    'AWS.s3.downloadFile.failedMany',
+                    'Failed downloads:\n{0}\nSee logs for full list of failed items',
+                    failedKeys.toString().split(',').slice(0, 5).join('\n')
+                ),
+                outputChannel
+            )
+        } else {
+            showOutputMessage(
+                localize(
+                    'AWS.s3.download.failed',
+                    'Failed downloads:\n{0}',
+                    failedKeys.toString().split(',').join('\n')
+                ),
+                outputChannel
+            )
+        }
+
+        const response = await window.showErrorMessage(
+            localize('AWS.s3.downLoad.retryPrompt', 'S3 Download: {0}/{1} failed.', failed.length, files.length),
+            localizedText.retry,
+            localizedText.skip
+        )
+
+        if (response === localizedText.retry) {
+            failed = await downloadAllFiles(failed, saveLocation, node.s3)
+        } else {
+            break
+        }
+    }
+}
+
+async function promptForSaveFolderLocation(window = Window.vscode()): Promise<vscode.Uri | undefined> {
+    const saveLocation = getDefaultDownloadPath()
+    const folderLocation = await window.showOpenDialog({
+        defaultUri: vscode.Uri.file(saveLocation),
+        openLabel: localize('AWS.s3.downloadFolder.openButton', 'Download Folder Here'),
+        canSelectFolders: true,
+        canSelectFiles: false,
+        canSelectMany: false,
+    })
+
+    if (!folderLocation || folderLocation.length == 0) {
+        return undefined
+    }
+
+    return folderLocation[0]
+}
+
+/**
+ * Download all S3 Files provided, divided in batches
+ *
+ * @param fileList
+ * @param saveLocation
+ * @param client
+ * @returns list of files that failed to download
+ */
+async function downloadAllFiles(fileList: S3File[], saveLocation: vscode.Uri, client: S3Client): Promise<S3File[]> {
```

Comment:
might be worth asserting that `saveLocation` is a folder

Reply:
They can only choose a folder from the prompt and we potentially might add one, but it wouldn't hurt to add a check.